project/apps/matching-service/src/matchEngine/matchEngine.service.ts

```diff
@@ -1,32 +1,84 @@
 import { MatchSupabase } from 'src/db/match.supabase';
 import { MatchingGateway } from 'src/matching.gateway';
+import { MatchDataDto, MatchRequestMsgDto } from '@repo/dtos/match';
+import { MATCH_TIMEOUT } from 'src/constants/queue';
+import { ClientProxy } from '@nestjs/microservices';
+import { firstValueFrom } from 'rxjs';
+import {QuestionCollectionDto, QuestionFiltersDto,
+} from '@repo/dtos/questions';
 
 @Injectable()
 export class MatchEngineService {
+  private readonly logger = new Logger(MatchEngineService.name);
   constructor(
+    @Inject('QUESTION_SERVICE') private readonly questionServiceClient: ClientProxy,
     private readonly matchEngineProduceExpiry: MatchExpiryProducer,
     private readonly matchRedis: MatchRedis,
     private readonly matchSupabase: MatchSupabase,
     private readonly matchGateway: MatchingGateway,
   ) {}
+  
+  /**
+   * Generates a match for the user based on the match request.
+   * If a match is found, the match data is saved to the database.
+   * If no match is found, the match request is added to the matching queue with an expiry time.
+   * @param matchRequest MQ message containing user id, array of categories and complexity to match
+   * @returns 
+   */
 
-  async generateMatch(matchRequest: any) {
-    // Performs some sort of redis checking
-    // Add to supabase if match found
-    // If no match found produce match expiry
-
-    // TODO: Replace with userId of user
-    this.matchGateway.sendMatchFound({
-      userId: '3821549f-001f-4fb9-beb8-8fb293747519',
-      message: 'success',
-    });
-
-    this.matchEngineProduceExpiry.enqueueMatchExpiryRequest(
-      matchRequest,
-      30000,
-    );
+  async generateMatch(matchRequest: MatchRequestMsgDto) {
+    const {userId, category, complexity} = matchRequest;
+
+    const matchedUser = await this.matchRedis.findPotentialMatch({category, complexity});
+
+    if (matchedUser) {
+      this.logger.log(`Match found for user ${userId} and ${matchedUser.userId}`);
+
+      const filters: QuestionFiltersDto = {
+        categories: category,
+        complexities: [complexity],
+      };
+
+      const selectedCollection = await firstValueFrom(
+        this.questionServiceClient.send<QuestionCollectionDto>({ cmd: 'get_questions' }, filters)
+      );
+      
+      if (!selectedCollection || !selectedCollection.questions) {
+        this.logger.error(`Match Failed. No questions found for category ${category} and complexity ${complexity}`);
+        return;
+      } 
+
+      // Choosing by random because get_questions will return collection in same sequence
+      const ind: number =
+      Math.floor(Math.random() * selectedCollection.questions.length);
+      
+      const matchData: MatchDataDto = {
+        user1_id: userId,
+        user2_id: matchedUser.userId,
+        complexity: complexity,
+        category: category,
+        id: matchedUser.matchId,
+        question_id: selectedCollection.questions[ind].id
+      };
```

Comment:
should this logic be abstracted to be done by supabase instead? meaning we call it in match.supabase.ts. can reference this [video](https://www.youtube.com/watch?v=ez8dfCYQGvE) for how to get a random row in supabase

Reply:
Will check that out. As for the match.supabase.ts part, do you mean for match.supabase.ts side to call/invoke the question repository func that will be responsible for doing the random selection from qn db?

Reply:
oh i meant like matchEngine.service will call matchSupabase.getRandom and let the logic of getting a random question be handled by matchSupabase. but since there is no function that supabase provides to get a random row, we have to either use the view method or create a function in the dashboard that we can call but the calling of this logic in either the view method or calling a function that we define on the supabase dashboard should be done by match.supabase